### DISCUSSION TITLE
Are there methods to build simdjson value/element in the code?

### DISCUSSION THREAD

**Discussion by bobjiang82** [simdjson/simdjson#1520] [Q&A]
We are evaluating replacing rapidjosn with simdjson in a project. In the code, we created a rapidjson::Value object to represent an array for printing later, as shown below. I read over simdjson APIs and don't find similar methods. Does this feature exist in simdjson? If not, is there a plan to add it?
```
        rapidjson::Value resultJson;
        rapidjson::Document::AllocatorType& a = sLastParsedJson.GetAllocator();
        resultJson.SetArray();
        for (auto doc : docArray)
        {
            resultJson.PushBack(rapidjson::Value(*doc, a), a);
        }
        resultJson.Accept(writer);
```

**Comment by lemire** [comment]
@bobjiang82 There is not a JSON builder interface at this time in simdjson. The simdjson library provides a parser. Serialization of your data structures in not currently in scope.

**Reply by bobjiang82** [reply]
Thanks for the reply.
I think simdjson::to_string() is used to serialize the existing element/value?
The background is we would like to support wildcard * in the middle of the path. e.g.
input json: [[1],[2],3,4,5]
the search path: /[*]/[0]
the expected output: [1, 2]
Our current implementation is to get the expected json element in each layer, then use a customized print_simdjson() function to output the '[', ',', ']', and each element (e.g. 1, 2).
If the json element '[1,2]' could be constructed in the code, we could use simdjson::to_string() to serialize it. That would be much more consistent and easy to maintain.

**Reply by lemire** [reply]
> I think simdjson::to_string() is used to serialize the existing element/value?

That's correct.

> If the json element '[1,2]' could be constructed in the code, we could use simdjson::to_string() to serialize it. That would be much more consistent and easy to maintain.

I understand.

If your goal is to start with an input JSON and filter it, then constructing and mutating an intermediary data structure is almost surely wasteful. We are promoting On Demand for such use cases except that as of 0.9.x, we are still lacking string serialization as part of On Demand. It is something we will no doubt soon add.

**Reply by bobjiang82** [reply]
Thanks for the reply and glad to know it will be added to On Demand API soon.
Just to double check, since we need to support wildcard * in the **middle**, it is not a simple filtering to get one element/array; but to get a list of elements in different locations. That's why we wanted to construct an intermediary array using these elements and then feed the array to to_string().
The current dom to_string() doesn't support this use case (* in the middle). Will the OnDemand string serialization function support it?

**Reply by lemire** [reply]
> That's why we wanted to construct an intermediary array using these elements and then feed the array to to_string().


Why not just serialize the numbers to a string buffer right away? Why would you build an intermediary data structure?

**Reply by bobjiang82** [reply]
Yeah, that may work as well. It requires some refactoring of our existing code.
Thanks!

**Reply by lemire** [reply]
@bobjiang82 The reason I make this proposal is that from a performance point of view, you usually (though not always) win by removing intermediate stages.
